unfoldingWord/translationCore
# Revamped fetch datas & loading process

## This pull request addresses:
- Removed fetch data from componentWillMount() & componentWillRecevceProps from tools.
- Made changes to improve loader
- Now setting the project bookName to the projectDetailReducer when the project is selected.
- Created a delay function helper to help delay dispatching actions.
- Changed all filesystem functions (fs) to use synchronous functions. For example, Now using outputJsonSync instead of outputJson. Now reading the Group Data directory synchronous instead of asynchronous making the read process much more efficient.
- Created the loadBiblesFromFS() action creator to handle loading bibles from the filesystem and loading them to the resources reducer. (this happens when the tools fetch data is not run.) Now loading bibles from filesystem when tools fetch data isn't run (loadBiblesFromFS).
- Added docs consistently throughout the codebase.
- Now flushing/clearing data much more cleanly than before when switching tools wich fixes multiple bugs when switching tools.
- Fixed bug where groupIndex was being overwritten in the filesystem by saving an empty array
- Moved setProjectPath, setProjectManifest actions to the appropriate action creator file (projectDetailsActions.js)

## How to test this pull request:
**Make sure you pull from both tools the branch with the name feature/mc-1499**
- [ ] Load a large project like Luke or Psalm and it should load.
- [ ] The Loader should work fine. although should not show progress since that feature was removed.
- [ ] you should be able to switch tools and project very quickly no matter how large the project is.
- [ ] Make sure to tests with a fresh project. or just make sure your project isn't corrupted. 

---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/unfoldingword-dev/translationcore/1602)
